src/crypto/muhash.cpp

```diff
@@ -0,0 +1,310 @@
+    }
+}
+
+void Multiply(Num3072& in_out, const Num3072& a)
+{
+    limb_t c0 = 0, c1 = 0;
+    Num3072 tmp;
+
+    /* Compute limbs 0..N-2 of in_out*a into tmp, including one reduction. */
+    for (int j = 0; j < LIMBS - 1; ++j) {
+        limb_t d0 = 0, d1 = 0, d2 = 0, c2 = 0;
+        mul(d0, d1, in_out.limbs[1 + j], a.limbs[LIMBS + j - (1 + j)]);
+        for (int i = 2 + j; i < LIMBS; ++i) muladd3(d0, d1, d2, in_out.limbs[i], a.limbs[LIMBS + j - i]);
+        mulnadd3(c0, c1, c2, d0, d1, d2, MAX_PRIME_DIFF);
+        for (int i = 0; i < j + 1; ++i) muladd3(c0, c1, c2, in_out.limbs[i], a.limbs[j - i]);
+        extract3(c0, c1, c2, tmp.limbs[j]);
+    }
+
+    /* Compute limb N-1 of a*b into tmp. */
+    limb_t c2 = 0;
+    for (int i = 0; i < LIMBS; ++i) muladd3(c0, c1, c2, in_out.limbs[i], a.limbs[LIMBS - 1 - i]);
+    extract3(c0, c1, c2, tmp.limbs[LIMBS - 1]);
+
+    /* Perform a second reduction. */
+    muln2(c0, c1, MAX_PRIME_DIFF);
+    for (int j = 0; j < LIMBS; ++j) {
+        add2(c0, c1, tmp.limbs[j]);
+        extract2(c0, c1, in_out.limbs[j]);
+    }
+#ifdef DEBUG
+    assert(c1 == 0);
+    assert(c0 == 0 || c0 == 1);
+#endif
+    /* Perform a potential third reduction. */
+    if (c0) FullReduce(in_out);
+}
+
+void Square(Num3072& in_out)
+{
+    limb_t c0 = 0, c1 = 0;
+    Num3072 tmp;
+
+    /* Compute limbs 0..N-2 of in_out*in_out into tmp, including one reduction. */
+    for (int j = 0; j < LIMBS - 1; ++j) {
+        limb_t d0 = 0, d1 = 0, d2 = 0, c2 = 0;
+        for (int i = 0; i < (LIMBS - 1 - j) / 2; ++i) muldbladd3(d0, d1, d2, in_out.limbs[i + j + 1], in_out.limbs[LIMBS - 1 - i]);
+        if ((j + 1) & 1) muladd3(d0, d1, d2, in_out.limbs[(LIMBS - 1 - j) / 2 + j + 1], in_out.limbs[LIMBS - 1 - (LIMBS - 1 - j) / 2]);
+        mulnadd3(c0, c1, c2, d0, d1, d2, MAX_PRIME_DIFF);
+        for (int i = 0; i < (j + 1) / 2; ++i) muldbladd3(c0, c1, c2, in_out.limbs[i], in_out.limbs[j - i]);
+        if ((j + 1) & 1) muladd3(c0, c1, c2, in_out.limbs[(j + 1) / 2], in_out.limbs[j - (j + 1) / 2]);
+        extract3(c0, c1, c2, tmp.limbs[j]);
+    }
+
+    limb_t c2 = 0;
+    for (int i = 0; i < LIMBS / 2; ++i) muldbladd3(c0, c1, c2, in_out.limbs[i], in_out.limbs[LIMBS - 1 - i]);
+    extract3(c0, c1, c2, tmp.limbs[LIMBS - 1]);
+
+    /* Perform a second reduction. */
+    muln2(c0, c1, MAX_PRIME_DIFF);
+    for (int j = 0; j < LIMBS; ++j) {
+        add2(c0, c1, tmp.limbs[j]);
+        extract2(c0, c1, in_out.limbs[j]);
+    }
+#ifdef DEBUG
+    assert(c1 == 0);
+    assert(c0 == 0 || c0 == 1);
+#endif
+    /* Perform a potential third reduction. */
+    if (c0) FullReduce(in_out);
+}
+
+void Inverse(Num3072& out, const Num3072& a)
+{
+    // For fast exponentiation a sliding window exponentiation with repunit
+    // precomputation is utilized. See "Fast Point Decompression for Standard
+    // Elliptic Curves" (Brumley, Järvinen, 2008).
+
+    Num3072 p[12]; // p[i] = a^(2^(2^i)-1)
+    Num3072 x;
```

Comment:
Why is this temporary needed? Can we just construct `out` here?

Changing the function signature to `Num3072 Inverse(const Num3072& in)` would use RVO and avoid copying `x` to `out`.

Reply:
Done